itest/litd_custom_channels_test.go

```diff
@@ -1675,3 +1678,207 @@ func testCustomChannelsBreach(_ context.Context, net *NetworkHarness,
+	fundingScriptKey := fundingScriptTree.TaprootKey
+	fundingScriptTreeBytes := fundingScriptKey.SerializeCompressed()
+
+	t.Logf("Minted %d lightning cents, syncing universes...", cents.Amount)
+	syncUniverses(t.t, charlieTap, dave)
+	t.Logf("Universes synced between all nodes, distributing assets...")
+
+	charlieBalance := cents.Amount
+
+	fundRespCD, err := charlieTap.FundChannel(
+		ctxb, &tchrpc.FundChannelRequest{
+			AssetAmount:        charlieBalance,
+			AssetId:            assetID,
+			PeerPubkey:         daveTap.node.PubKey[:],
+			FeeRateSatPerVbyte: 5,
+			PushSat:            0,
+		},
+	)
+	require.NoError(t.t, err)
+	t.Logf("Funded channel between Charlie and Dave: %v", fundRespCD)
+
+	// Make sure the pending channel shows up in the list and has the
+	// custom records set as JSON.
+	assertPendingChannels(
+		t.t, charlieTap.node, assetID, 1, charlieBalance, 0,
+	)
+
+	// Let's confirm the channel.
+	mineBlocks(t, net, 6, 1)
+
+	assertAssetBalance(t.t, charlieTap, assetID, cents.Amount)
+
+	// There should only be a single asset piece for Charlie, the one in the
+	// channel.
+	assertNumAssetOutputs(t.t, charlieTap, assetID, 1)
+	assertAssetExists(
+		t.t, charlieTap, assetID, charlieBalance,
+		fundingScriptKey, false, true, true,
+	)
+
+	// Assert that the proofs for both channels has been uploaded to the
+	// designated Universe server.
+	assertUniverseProofExists(
+		t.t, universeTap, assetID, groupKey, fundingScriptTreeBytes,
+		fmt.Sprintf("%v:%v", fundRespCD.Txid, fundRespCD.OutputIndex),
+	)
+
+	// Make sure the channel shows the correct asset information.
+	assertAssetChan(
+		t.t, charlieTap.node, daveTap.node, charlieBalance, assetID,
+	)
+
+	logBalance(t.t, nodes, assetID, "initial")
+
+	// Normal case.
+	// Send 50 assets from Charlie to Dave.
+	sendAssetKeySendPayment(
+		t.t, charlie, dave, 50, assetID,
+		fn.None[int64](), lnrpc.Payment_SUCCEEDED,
+	)
+
+	logBalance(t.t, nodes, assetID, "after 50 assets")
+
+	// Normal case.
+	// Send 1k sats from Charlie to Dave.
+	sendKeySendPayment(t.t, charlie, dave, 1000)
+
+	logBalance(t.t, nodes, assetID, "after 1k sats")
+
+	// Edge case: The channel reserve check should trigger, and we should
+	// get a payment failure, not a timeout.
+	//
+	// Now Dave tries to send 50 assets to Charlie. There shouldn't be
+	// enough sats in the channel.
+
+	done := make(chan bool, 1)
+	// Assume an acceptable completion window which is half the payment
+	// timeout. If the payment succeeds within this duration this means we
+	// didn't fall into a routing loop.
+	timeoutChan := time.After((time.Second * PaymentTimeoutSec) / 2)
+
+	go func() {
+		sendAssetKeySendPayment(
+			t.t, dave, charlie, 50, assetID,
+			fn.None[int64](), lnrpc.Payment_FAILED,
+		)
+
+		done <- true
+	}()
+
+	select {
+	case <-done:
+	case <-timeoutChan:
+		t.Fatalf("Payment didn't fail within expected time duration")
+
+	}
+
+	logBalance(t.t, nodes, assetID, "after failed 50 assets")
```

Comment:
Would be nice if we could also assert the payment's failure reason, that it's actually the `INSUFFICIENT_BALANCE` code. I think we return that somewhere.

Reply:
yeap good idea, it's already available here through the `SendPayment` RPC response, `result.FailureReason`

Reply:
Since this is meant to fail in the `paymentLifecycle` phase, we should get a `NO_ROUTE` instead

The reason the failure happens there is because that's when we actually provide the `500 sats` amount to the `PaymentBandwidth` hook. See more on [this comment](https://github.com/GeorgeTsagk/taproot-assets/blob/tap-chan-liquidity-fixes/tapchannel/aux_traffic_shaper.go#L146-L163)